### PR TITLE
[AIRFLOW-986] HiveCliHook ignores 'proxy_user' value in a connection's extra parameter

### PR DESCRIPTION
**Pull request by OmerJog**
2nd attempt to https://issues.apache.org/jira/browse/AIRFLOW-986
This PR is based on previous stale PR https://github.com/apache/airflow/pull/2153 which was not  merged nor reviewed

### CODE REVIEW

File: airflow/hooks/hive_hooks.py [hunk 1]
```diff
@@ -122,11 +122,13 @@ def _prepare_cli_cmd(self):
                     template = utils.replace_hostname_pattern(
                         utils.get_components(template))
 
-                proxy_user = ""  # noqa
+                proxy_user = conn.extra_dejson.get('proxy_user', "")  # noqa
                 if conn.extra_dejson.get('proxy_user') == "login" and conn.login:
```
Comment (feluelle) [inline airflow/hooks/hive_hooks.py]:
```suggestion
                if proxy_user == "login" and conn.login:
```

File: airflow/hooks/hive_hooks.py [hunk 2]
```diff
@@ -122,11 +122,13 @@ def _prepare_cli_cmd(self):
...
                     proxy_user = "hive.server2.proxy.user={0}".format(conn.login)
                 elif conn.extra_dejson.get('proxy_user') == "owner" and self.run_as:
```
Comment (feluelle) [inline airflow/hooks/hive_hooks.py]:
```suggestion
                 elif proxy_user == "owner" and self.run_as:
```

File: airflow/hooks/hive_hooks.py (hunk 1 above)
Comment (feluelle) [inline airflow/hooks/hive_hooks.py]:
```suggestion
                 if proxy_user == "login" and conn.login:
```

**Review by feluelle — APPROVED** [review]
LGTM besides a few suggestions 👍

**Comment by OmerJog** [comment]
@feluelle PTAL

File: airflow/hooks/hive_hooks.py [hunk 3]
```diff
@@ -122,11 +122,13 @@ def _prepare_cli_cmd(self):
                     template = utils.replace_hostname_pattern(
                         utils.get_components(template))
 
-                proxy_user = ""  # noqa
-                if conn.extra_dejson.get('proxy_user') == "login" and conn.login:
+                proxy_user = conn.extra_dejson.get('proxy_user', "")  # noqa
```
Comment (feluelle) [inline airflow/hooks/hive_hooks.py]: actually.. sorry but what do you think of a [_factory_](https://en.wikipedia.org/wiki/Factory_method_pattern) for the `proxy_user` like:
```
def _get_proxy_user(self):
    proxy_user_value = conn.extra_dejson.get('proxy_user', "")

    if proxy_user_value == "login" and conn.login:
        return "hive.server2.proxy.user={login}".format(login=conn.login)
    if proxy_user_value == "owner" and self.run_as:
        return "hive.server2.proxy.user={run_as}".format(run_as=self.run_as)
    if proxy_user_value != "":  # There is a custom proxy user
        return "hive.server2.proxy.user={proxy_user_value}".format(proxy_user_value=proxy_user_value)

     return proxy_user_value  # The default proxy user (undefined)
```
and then call `proxy_user = self._get_proxy_user()`

Comment (OmerJog) [reply airflow/hooks/hive_hooks.py]: @feluelle Sure. Done.

**Review by feluelle — APPROVED** [review]
Nice. LGTM 👍

**Comment by OmerJog** [comment]
@Fokko  PTAL

File: airflow/hooks/hive_hooks.py [hunk 4]
```diff
@@ -103,6 +103,21 @@ def __init__(
         self.mapred_queue_priority = mapred_queue_priority
         self.mapred_job_name = mapred_job_name
 
+    def _get_proxy_user(self):
+        """
+        This function set the proper proxy_user value in case the user overwtire the default.
+        """
+        conn = self.conn
+
+        proxy_user_value = conn.extra_dejson.get('proxy_user', "")  # noqa
```
Comment (Fokko) [inline airflow/hooks/hive_hooks.py]: Do we still need this `# noqa`?

**Review by Fokko — APPROVED** [review]
One minor nit, apart from that LGTM. Sorry for taking so long to review this @OmerJog

**Comment by Fokko** [comment]
Thanks @OmerJog

Let us wait for the CI to become green.

**Comment by OmerJog** [comment]
@Fokko  CI is green :)

**Comment by OmerJog** [comment]
@Fokko ping :)
Also this can be cherry picked to 1.10.4 ?

**Comment by Fokko** [comment]
Done, and cherry-picked for 1.10.4 👍 Thanks @OmerJog

**Pull request merged**